src/lib.rs

```diff
@@ -0,0 +1,431 @@
+    pub source: ReleaseSource,
+    pub version: String,
+}
+
+#[cfg(feature = "github_releases")]
+pub fn get_github_releases(name: &String, owner: &String) -> AxoupdateResult<Vec<Release>> {
+    let client = reqwest::blocking::Client::new();
+    let resp: Vec<Release> = client
+        .get(format!("{GITHUB_API}/repos/{owner}/{name}/releases"))
+        .header(ACCEPT, "application/json")
+        .header(
+            USER_AGENT,
+            format!("axoupdate/{}", env!("CARGO_PKG_VERSION")),
+        )
+        .send()?
+        .json()?;
+
+    Ok(resp)
+}
+
+#[cfg(feature = "axo_releases")]
+pub fn get_axo_releases(name: &String, owner: &String) -> AxoupdateResult<Vec<Release>> {
+    let abyss = Gazenot::new_unauthed("github".to_string(), owner)?;
+    let release_lists = tokio::runtime::Builder::new_current_thread()
+        .worker_threads(1)
+        .max_blocking_threads(128)
+        .enable_all()
+        .build()
+        .expect("Initializing tokio runtime failed")
+        .block_on(abyss.list_releases_many(vec![name.to_owned()]))?;
+    let Some(our_release) = release_lists.iter().find(|rl| &rl.package_name == name) else {
+        return Err(AxoupdateError::ReleaseNotFound {
+            app_name: name.to_owned(),
+        });
+    };
+
+    Ok(our_release
+        .releases
+        .iter()
+        .map(Release::from_gazenot)
+        .collect())
+}
+
+pub fn get_latest_stable_release(
+    name: &String,
+    owner: &String,
+    release_type: &ReleaseSourceType,
+) -> AxoupdateResult<Option<Release>> {
+    let releases = match release_type {
+        #[cfg(feature = "github_releases")]
+        ReleaseSourceType::GitHub => get_github_releases(name, owner)?,
+        #[cfg(not(feature = "github_releases"))]
+        ReleaseSourceType::GitHub => {
+            return Err(AxoupdateError::BackendDisabled {
+                backend: "github".to_owned(),
+            })
+        }
+        #[cfg(feature = "axo_releases")]
+        ReleaseSourceType::Axo => get_axo_releases(name, owner)?,
+        #[cfg(not(feature = "axo_releases"))]
+        ReleaseSourceType::Axo => {
+            return Err(AxoupdateError::BackendDisabled {
+                backend: "axodotdev".to_owned(),
+            })
+        }
+    };
+
+    Ok(releases.into_iter().find(|r| !r.prerelease))
+}
+
+pub fn get_app_name() -> Option<String> {
+    if cfg!(debug_assertions) {
+        Some("cargo-dist".to_owned())
+    } else if let Some(path) = args().next() {
+        Utf8PathBuf::from(&path)
+            .file_name()
+            .map(|s| s.strip_suffix(".exe").unwrap_or(s))
+            .map(|s| s.strip_suffix("-update").unwrap_or(s))
+            .map(|s| s.to_owned())
+    } else {
+        None
+    }
+}
+
+pub fn get_config_path(app_name: &String) -> AxoupdateResult<Utf8PathBuf> {
+    if cfg!(debug_assertions) {
+        Ok(Utf8PathBuf::try_from(current_dir()?)?)
+    } else {
+        let Some(home) = homedir::get_my_home()? else {
```

Comment:
specifically https://docs.rs/dirs/latest/dirs/fn.config_local_dir.html

Reply:
I started with `dirs`, but had to back out because I couldn't write the receipts using it. dirs only lets you access dir info for the platform it's running on; you can't, say, ask for what Windows dirs should look like when you're running on Linux. As a result, I could only really assemble actual receipt paths in bash/powershell, where I didn't have access to dirs. Here I'm matching exactly the logic I used there, rather than dirs's special logic that might not match it 1:1.